tests/test_syntax/inline/test_autolinks.py

```diff
@@ -0,0 +1,55 @@
+"""
+Python Markdown
+
+A Python implementation of John Gruber's Markdown.
+
+Documentation: https://python-markdown.github.io/
+GitHub: https://github.com/Python-Markdown/markdown/
+PyPI: https://pypi.org/project/Markdown/
+
+Started by Manfred Stienstra (http://www.dwerg.net/).
+Maintained for a few years by Yuri Takhteyev (http://www.freewisdom.org).
+Currently maintained by Waylan Limberg (https://github.com/waylan),
+Dmitry Shachnev (https://github.com/mitya57) and Isaac Muse (https://github.com/facelessuser).
+
+Copyright 2007-2018 The Python Markdown Project (v. 1.7 and later)
+Copyright 2004, 2005, 2006 Yuri Takhteyev (v. 0.2-1.6b)
+Copyright 2004 Manfred Stienstra (the original version)
+
+License: BSD (see LICENSE.md for details).
+"""
+
+from markdown.test_tools import TestCase
+
+
+class TestAutomaticLinks(TestCase):
+
+    def test_email_address(self):
+        self.assertMarkdownRenders(
+            """asdfasdfadsfasd <yuri@freewisdom.org> or you can say """,
+            """<p>asdfasdfadsfasd <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#121;&#117;&#114;&#105;&#64;&#102;&#114;&#101;&#101;&#119;&#105;&#115;&#100;&#111;&#109;&#46;&#111;&#114;&#103;">&#121;&#117;&#114;&#105;&#64;&#102;&#114;&#101;&#101;&#119;&#105;&#115;&#100;&#111;&#109;&#46;&#111;&#114;&#103;</a> or you can say </p>"""
+        )
+
+    def test_mailto_email_address(self):
+        self.assertMarkdownRenders(
+            """instead <mailto:yuri@freewisdom.org>""",
+            """<p>instead <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#121;&#117;&#114;&#105;&#64;&#102;&#114;&#101;&#101;&#119;&#105;&#115;&#100;&#111;&#109;&#46;&#111;&#114;&#103;">&#121;&#117;&#114;&#105;&#64;&#102;&#114;&#101;&#101;&#119;&#105;&#115;&#100;&#111;&#109;&#46;&#111;&#114;&#103;</a></p>"""
+        )
+
+    def test_email_address_with_ampersand(self):
+        self.assertMarkdownRenders(
+            """<bob&sue@example.com>""",
+            """<p><a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#98;&#111;&#98;&#38;&#115;&#117;&#101;&#64;&#101;&#120;&#97;&#109;&#112;&#108;&#101;&#46;&#99;&#111;&#109;">&#98;&#111;&#98;&amp;&#115;&#117;&#101;&#64;&#101;&#120;&#97;&#109;&#112;&#108;&#101;&#46;&#99;&#111;&#109;</a></p>"""
+        )
+
+    def test_invalid_email_address_local_part(self):
+        self.assertMarkdownRenders(
+            """Missing local-part <@domain>""",
+            """<p>Missing local-part &lt;@domain&gt;</p>"""
+        )
+
+    def test_invalid_email_address_domain(self):
+        self.assertMarkdownRenders(
+            """Missing domain <local-part@>""",
+            """<p>Missing domain <local-part@></p>"""
+        )
```

Comment:
The different behavior still bothers me so I checked [Babelmark][1] and this matches the reference implementation. Although, I suspect that is a bug. In fact, no other implementation has this behavior.

I suspect the issue is elsewhere in the code and could be classified as a separate bug. However, I'm not comfortable codifying incorrect behavior in a test, so we either need to fix the issue here or remove this test. I'd rather fix the issue.

[1]: https://johnmacfarlane.net/babelmark2/?normalize=1&text=%3C%40domain%3E%0A%0A%3Clocal-part%40%3E

Reply:
I see the issue. Note the inline HTML regex:

https://github.com/Python-Markdown/markdown/blob/c1724406ae518a4b295406cdb076a5804fee3ef3/markdown/inlinepatterns.py#L166

That correctly requires the first character after the opening bracket to be a letter (`A-Za-z`) or a slash (`/`). Of course, `<@domain>` fails that test and is not treated an raw HTML and gets escaped. However, as anything other than angle brackets are accepted for any character after the first character, then `<local-part@>` is treated as raw HTML and not escaped. And we can't just insert a `@` in the disallowed characters (`[^<>@]`) as this is valid HTML: `<a href="mailto:email@example.com">email</a>`. Of course, and `@` outside of quotes could probably be disallowed, but that would require a much more sophisticated regex.

Outside of an easy fix, we may just leave it as-is. Thoughts anyone?

Reply:
A simple solution may be to disallow some characters before a space character, but allow anything after an optional space. Then we don't need to get into the specifics of whether things are wrapped in quotes or not. We would simply be verifying that the tag name contains only valid characters. I don't see any reason to be concerned over whether any attributes are valid or not.

Perhaps something like this: `\/?[a-zA-Z][^<>@ ]*( [^<>]*)?`.

Note that I also moved the optional opening slash out of the first character That way we are still confirming that the first character after the opening slash is alpha only. For reference see the [HTML spec][1]. Of course, [XML][2] is more complicated, but XHTML would be the intersection of the two, so I think we are okay here. 

We could probably make some additional changes, but that is another issue. The above should address this issue alone. That is, if it doesn't break any of our existing tests.

[1]: https://html.spec.whatwg.org/multipage/syntax.html#syntax-tag-name
[2]: https://www.w3.org/TR/xml/#NT-S

Reply:
> A simple solution may be to disallow some characters before a space character, but allow anything after an optional space.

This sounds like the best thing to do.